File: DotDice/ParserCombinators.cs
using System.Diagnostics.Tracing;

namespace DotDice;

public static class ParserCombinators
{
    // Basic combinators

    // Return a parser that always succeeds with the given value
    public static Parser<T> Return<T>(T value)
    {
        return (input, position) => ParserResult<T>.Succeed(value, position);
    }

    // Run the parser and return the result if it succeeds
    public static ParserResult<T> Run<T>(this Parser<T> parser, string input)
    {
        return parser(input, 0);
    }

    // Choice combinator: try the first parser, if it fails, try the second
    public static Parser<T> Or<T>(this Parser<T> p1, Parser<T> p2)
    {
        return (input, position) =>
        {
            var result1 = p1(input, position);
            if (result1.Success)
            {
                return result1;
            }
            else
            {
                return p2(input, position);
            }
        };
    }

    // Apply a function to the result of a parser
    public static Parser<U> Select<T, U>(this Parser<T> parser, Func<T, U> selector)
    {
        return (input, position) =>
        {
            var result = parser(input, position);
            if (result.Success)
            {
                return ParserResult<U>.Succeed(selector(result.Value), result.Position);
            }
            else
            {
                return ParserResult<U>.Fail(result.Error, result.Position);
            }
        };
    }

    // Sequence two parsers and combine their results with a function
    public static Parser<V> SelectMany<T, U, V>(
        this Parser<T> parser,
        Func<T, Parser<U>> selector,
        Func<T, U, V> projector)
    {
        return (input, position) =>
        {
            var result1 = parser(input, position);
            if (!result1.Success)
            {
                return ParserResult<V>.Fail(result1.Error, result1.Position);
            }

            var result2 = selector(result1.Value)(input, result1.Position);
            if (!result2.Success)
            {
                return ParserResult<V>.Fail(result2.Error, result2.Position);
            }

            return ParserResult<V>.Succeed(projector(result1.Value, result2.Value), result2.Position);
        };
    }

    // Other helpful combinators

    // Parse a single character satisfying a predicate
    public static Parser<char> Satisfy(Predicate<char> predicate)
    {
        return (input, position) =>
        {
            if (position >= input.Length)
            {
                return ParserResult<char>.Fail("Unexpected end of input", position);
            }

            char c = input[position];
            if (predicate(c))
            {
                return ParserResult<char>.Succeed(c, position + 1);
            }
            else
            {
                return ParserResult<char>.Fail($"Unexpected character: '{c}'", position);
            }
        };
    }

    // Parse a specific character
    public static Parser<char> Char(char c)
    {
        return Satisfy(ch => ch == c);
    }

    // Parse any character in a string
    public static Parser<char> AnyOf(string chars)
    {
        return Satisfy(chars.Contains);
    }

    // Parse one or more occurrences of a parser
    public static Parser<List<T>> OneOrMore<T>(this Parser<T> parser)
    {
        return parser.SelectMany(
            x => parser.Many().Select(xs => new List<T> { x }.Concat(xs).ToList()),
            (x, xs) => xs);
    }

    // Parse zero or more occurrences of a parser
    public static Parser<List<T>> Many<T>(this Parser<T> parser)
    {
        return parser.OneOrMore().Or(Return(new List<T>()));
    }

    // Parse a string
    public static Parser<string> String(string str)
    {
        return str.Select(Char)
            .Aggregate((p1, p2) => p1.Then(p2))
            .Select(c => c.ToString());
    }

    // Optionally parse something
    public static Parser<Option<T>> Optional<T>(this Parser<T> parser)
    {
        return parser.Select(x => new Option<T>(x))
            .Or(Return(Option<T>.None()));
    }

    // A parser for whitespace
    public static readonly Parser<string> Whitespace =
        Satisfy(char.IsWhiteSpace)
            .Many()
            .Select(x => new string(x.ToArray()));

    // Skip whitespace before and after a parser
    public static Parser<T> Token<T>(this Parser<T> parser)
    {
        return Whitespace.Then(parser).ThenSkip(Whitespace);
    }

    // Sequence two parsers and keep the result of the first
    public static Parser<T> ThenSkip<T, U>(this Parser<T> parser, Parser<U> other)
    {
        return parser.SelectMany(_ => other, (x, _) => x);
    }

    // Sequence two parsers and keep the result of the second
    public static Parser<U> Then<T, U>(this Parser<T> parser, Parser<U> other)
    {
        return parser.SelectMany(_ => other, (_, y) => y);
    }

    // A parser for an integer with an optional negative sign
    public static readonly Parser<int> Integer =
        Char('-')
            .Optional()
            .Then(Satisfy(char.IsDigit).OneOrMore())
            .Select(
                result =>
                {
                    Option<char> sign = Option<char>.None();
                    List<char> digits;
                    if(result.First() == '-'){
                        sign = new('-');
                        digits = result.Skip(1)
                            .ToList();
                    } else {
                        digits = result;
                    }

                    var value = int.Parse(new string(digits.ToArray()));
                    return sign.HasValue ? -value : value;
                });

    // A parser for an optional integer
    public static readonly Parser<Option<int>> OptionalInteger = Integer.Optional();
}

File: DotDiceF/Core/ParserCombinators.fs
namespace DotDiceF.Core

open DotDiceF.Core
open System

/// A parser is a function that takes a string as input and returns a ParserResult.
type Parser<'T> = string -> ParserResult<'T>

module ParserCombinators =

    /// Basic Combinators
    /// 
    /// Returns a parser that always succeeds with the given value, without consuming any input.
    let pReturn x : Parser<'T> =
        fun input -> Success (x, input)

    /// Runs the parser and returns the result if it succeeds, otherwise fail.
    let runP parser input =
        parser input

    /// Choice combinator: tries the first parser, if it fails, tries the second.
    let ( <|> ) p1 p2 input =
        match p1 input with
        | Success _ as success -> success
        | Failure _ -> p2 input

    /// Apply a function to the result of a parser.
    let ( <!> ) f p input =
        match p input with
        | Success (x, rest) -> Success (f x, rest)
        | Failure err -> Failure err

    /// Sequence two parsers and combine their results.
    let ( <*> ) pF pX input =
        match pF input with
        | Failure err -> Failure err
        | Success (f, rest) ->
            match pX rest with
            | Failure err -> Failure err
            | Success (x, rest') -> Success (f x, rest')

    /// Sequence two parsers and keep the result of the first.
    let ( .>> ) p1 p2 input =
        match p1 input with
        | Failure err -> Failure err
        | Success (x, rest) ->
            match p2 rest with
            | Failure err -> Failure err
            | Success (_, rest') -> Success(x, rest')

    /// Sequence two parsers and keep the result of the second.
    let ( >>. ) p1 p2 input =
        match p1 input with
        | Failure err -> Failure err
        | Success (_, rest) -> p2 rest

    /// Other Helpful Combinators

    /// Parse a single character satisfying a predicate.
    let satisfy pred : Parser<char> =
        fun input ->
            if String.IsNullOrEmpty(input) then
                Failure "Unexpected end of input"
            else
                let c = input.[0]
                if pred c then
                    Success (c, input.[1..])
                else
                    Failure (sprintf "Unexpected character: '%c'" c)

    /// Parse a specific character.
    let pChar c : Parser<char> =
        satisfy ((=) c)

    /// Parse any character in a string.
    let anyOf chars = chars |> Seq.map pChar |> Seq.reduce (<|>)

    /// Parse one or more occurrences of a parser.
    let rec many1 p =
        (fun x xs -> x :: xs) <!> p <*> many p

    /// Parse zero or more occurrences of a parser.
    and many p =
        many1 p <|> pReturn []

    /// Parse a string.
    let pString str =
        str
        |> Seq.map pChar
        |> Seq.fold (fun acc p ->
            (fun accStr c -> accStr + string c) <!> acc <*> p) (pReturn "")

    /// Optionally parse something.
    let opt (p: Parser<'a>) : Parser<'a option> =
        let some = Some <!> p
        let none = pReturn None
        some <|> none

    /// A parser for whitespace.
     // A parser for whitespace.
    let whitespace: Parser<string> =
        many (satisfy Char.IsWhiteSpace) |> (fun p ->  (fun cl -> String.Concat(cl|> List.map string)) <!> p)

    /// Skip whitespace before and after a parser.
    let token (p: Parser<string>) : Parser<string> =
        whitespace >>. (p .>> whitespace)

    /// A parser for an integer with an optional negative sign.
    let pInteger : Parser<int> =
        let sign = opt (pChar '-')
        let digits = many1 (satisfy Char.IsDigit)
        (fun s (ds: char list) ->
            let str = String.Concat ds
            let n = Int32.Parse str
            match s with
            | Some '-' -> -n
            | _ -> n)
        <!> sign <*> digits

    /// A parser for an optional integer.
    let pOptionalInteger : Parser<int option> =
        opt pInteger